sys/net/gnrc/application_layer/tftp/gnrc_tftp.c

```diff
@@ -405,11 +405,14 @@ int gnrc_tftp_server(tftp_data_cb_t data_cb, tftp_start_cb_t start_cb, tftp_stop
     }
 
     /* context will be initialized when a connection is established */
-    tftp_context_t ctxt;
-    ctxt.data_cb = data_cb;
-    ctxt.start_cb = start_cb;
-    ctxt.stop_cb = stop_cb;
-    ctxt.enable_options = use_options;
+    tftp_context_t ctxt = {
+        .dst_port = GNRC_TFTP_DEFAULT_DST_PORT,
+        .src_port = GNRC_TFTP_DEFAULT_DST_PORT,
```

Comment:
I'm not sure this should be `SRC_PORT`... I'd like to do some personal testing for the "normal" operation first.

Reply:
Please let me how to confirm then in the test procedure (or ping someone more knowledgeable).

Reply:
Since the contributor that provided this is not active anymore, I'd say follow the README. Maybe also @bergzand has some input.

Reply:
Assuming my understanding of the protocol and the implementation here is sufficient, this `.src_port` is the UDP port used as source by RIOT and used as destination by the peer. Both *should* be ephermal ports specific for this connection.

Does it make sense to default it to something invalid/unconfigured? (Do we have something like that in gnrc?)

Reply:
> Assuming my understanding of the protocol and the implementation here is sufficient, this .src_port is the UDP port used as source by RIOT and used as destination by the peer. Both should be ephermal ports specific for this connection.

We are in server context here, so `src_port` should at least be `GNRC_TFTP_DEFAULT_DST_PORT` (which is the default TFTP port). Ephemeral ports in GNRC are part of the `gnrc_sock` implementation, which this implementation does not use :-/

> Does it make sense to default it to something invalid/unconfigured? (Do we have something like that in gnrc?)

If I leave it unconfigured, the testing procedure in #11772 runs into an assertion here

https://github.com/RIOT-OS/RIOT/blob/a01e70a4ab0050331b36ad59946637c62c48ed3a/sys/net/gnrc/transport_layer/udp/gnrc_udp.c#L283-L286

from the call here

https://github.com/RIOT-OS/RIOT/blob/a01e70a4ab0050331b36ad59946637c62c48ed3a/sys/net/gnrc/application_layer/tftp/gnrc_tftp.c#L967-L969

This is the only application protocol purely based on GNRC (and as soon as we have async sock, I prefer to have it ported to `sock_udp`, but first we need to fix the bug ;-)).

Reply:
Ephemeral port selection is included in the module btw:

https://github.com/RIOT-OS/RIOT/blob/a01e70a4ab0050331b36ad59946637c62c48ed3a/sys/net/gnrc/application_layer/tftp/gnrc_tftp.c#L368-L370

Problem is: it runs also, when the _server_ receives a valid message, causing this https://github.com/RIOT-OS/RIOT/pull/11773#issuecomment-508068123

Reply:
> If I leave it unconfigured, the testing procedure in #11772 runs into an assertion here […]

You did leave it unconfigered in your PR, right? Because doesn't this allow someone to crash a RIOT node running `gnrc_tftp` with `DEVELHELP = 1` by sending a crafted input?

Reply:
Arghs, this must have gotten lost, when I squashed :(

Reply:
Wait no, right. I moved the initialization to 2fce0950bf0d03a2702bba7312fa55119e70a085, as `_tftp_state_processes` relied on certain configurations of `ctxt->dst_port` I was unsure about. I tested it with your crafted input, and it did not crash (the server was not responding anymore, because a legal input changed its port...., see #11778).

Reply:
> I tested it with your crafted input, and it did not crash

My previous test input gets rejected due to the blocknum contained in it. If you compile with `ENABLE_DEBUG` you get (among other things):

```
tftp: not the packet we were waiting for, expected 1, received 0
```

If you change the blocknum in the test input `gnrc_tftp` crashes:

```
echo AAMAAWZvb2Jhcg== | base64 -d | nc -u '[ip-address%tap0]' 69
``

Can you reproduces this or am I simply doing something wrong?

Reply:
Yepp... Why did I touch this module -.- It's good that I deprecated it.